In [1]:
import pandas as pd
import pandas as pd
from sqlalchemy import create_engine
%load_ext sql

In [2]:
engine=create_engine('postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera')
connection=engine.connect()
print("Connected:", connection.closed == False)

Connected: True


In [3]:
%sql postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera

'Connected: postgres@medika_sejahtera'

# 1. Product & Reason Analysis

In [4]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT (*) AS total_record
FROM
    return_cleaned
GROUP BY
    return_reason_cleaned
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_record
Salah Input Sistem,28
Salah Pesan Customer,20
Barang Rusak,7
Kadaluarsa,4


In [27]:
%%sql
WITH total AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.return_reason_cleaned,
    COUNT(*) AS total_record,
    tl.total_records AS total_data,
    ROUND((COUNT(*)::numeric / NULLIF(tl.total_records,0)::numeric) * 100,2) AS rate
FROM
    return_cleaned rc
CROSS JOIN
    total tl
GROUP BY
    rc.return_reason_cleaned,
    total_data
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_record,total_data,rate
Salah Input Sistem,28,59,47.46
Salah Pesan Customer,20,59,33.90
Barang Rusak,7,59,11.86
Kadaluarsa,4,59,6.78


**Findings**  
Finding: System Input Error is the most common return reason (28 of 59 cases, 47.46%), followed by Customer Order Error (20 cases, 33.90%), Damaged Goods (7 cases, 11.86%), and Expired (4 cases, 6.78%).

In [ ]:
%%sql

SELECT
    CASE
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    COUNT(*) AS total_record
FROM 
    return_cleaned
WHERE 
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY 
    quarter
ORDER BY 
    quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_record
Q1,2
Q2,10
Q3,6
Q4,10


In [39]:
%%sql

WITH qr AS (
    SELECT
    CASE
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter_filter,
    COUNT(*) AS total_record_filter
    FROM 
        return_cleaned
    WHERE 
        return_reason_cleaned = 'Salah Input Sistem'
    GROUP BY 
        quarter_filter
)


SELECT
    CASE
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    qr.total_record_filter,
    COUNT(*) AS total_record,
    ROUND((NULLIF(qr.total_record_filter,0) * 1.0 / NULLIF(COUNT(*), 0))*100, 2) AS rate
FROM 
    return_cleaned rc
LEFT JOIN qr
    ON CASE
        WHEN EXTRACT(MONTH FROM rc.return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM rc.return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM rc.return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM rc.return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END = qr.quarter_filter
GROUP BY 
    quarter,
    qr.total_record_filter
ORDER BY 
    quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_record_filter,total_record,rate
Q1,2,8,25.00
Q2,10,15,66.67
Q3,6,14,42.86
Q4,10,22,45.45


**Findings**  
Among the quarters reliable for comparison (Q2, Q3, Q4 — Q1 excluded due to observed and expected counts below 5), the proportion of returns due to System Input Error is notably higher in Q2 (66.67%, observed 10 vs. expected 7.12; +19.21 points above the 47.46% baseline). Q3 and Q4 remain close to baseline (42.86% and 45.45%, respectively).

In [17]:
%%sql

SELECT
    region_cleaned,
    COUNT (*) AS total_record
FROM
    return_cleaned
WHERE
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY
    region_cleaned
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_record
Jakarta Barat,8
Jakarta Selatan,5
Bekasi,5
Jakarta Pusat,3
Bogor,2
Jakarta Timur,2
Depok,1
Jakarta Utara,1
Tangerang,1


In [38]:
%%sql

WITH rrc AS (
    SELECT
        region_cleaned,
        COUNT(*) AS total_record_filter
    FROM
        return_cleaned
    WHERE
        return_reason_cleaned = 'Salah Input Sistem'
    GROUP BY
        region_cleaned
)
SELECT
    rc.region_cleaned,
    rrc.total_record_filter,
    COUNT(*) AS total_record_all,
    ROUND( (COALESCE(rrc.total_record_filter,0) * 1.0 / NULLIF(COUNT(*),0) * 100)::numeric,2) AS rate
FROM
    return_cleaned rc
LEFT JOIN rrc
    ON rc.region_cleaned = rrc.region_cleaned
GROUP BY
    rc.region_cleaned,
    rrc.total_record_filter
ORDER BY
    rrc.total_record_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_record_filter,total_record_all,rate
Jakarta Barat,8,13,61.54
Bekasi,5,10,50.00
Jakarta Selatan,5,10,50.00
Jakarta Pusat,3,3,100.00
Bogor,2,4,50.00
Jakarta Timur,2,6,33.33
Depok,1,3,33.33
Jakarta Utara,1,5,20.00
Tangerang,1,5,20.00


**Findings**  
Of the 9 regions, only Jakarta Barat has an expected count large enough for a reliable comparison (6.17, out of 13 total cases). The result shows a slightly higher proportion of System Input Error returns in Jakarta Barat (61.54%, observed 8) compared to the 47.46% baseline. The remaining regions were excluded due to observed and/or expected counts below 5.

In [41]:
%%sql

SELECT
    kategori,
    COUNT (*) AS total_record
FROM
    return_cleaned
WHERE
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY
   kategori
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_record
Diagnostik,9
Consumable,6
Bedah & Tindakan,5
Laboratorium,5
Perawatan & Rawat Inap,3


In [43]:
%%sql

WITH rk AS (
    SELECT
        kategori,
        COUNT (*) AS total_record
    FROM
        return_cleaned
    WHERE
        return_reason_cleaned = 'Salah Input Sistem'
    GROUP BY
    kategori
)


SELECT
    rc.kategori,
    rk.total_record AS total_record_filter,
    COUNT (*) AS total_record,
    ROUND((NULLIF(rk.total_record,0) * 1.0 / NULLIF(COUNT(*), 0))*100, 2) AS rate
FROM
    return_cleaned rc
LEFT JOIN rk
    ON rc.kategori = rk.kategori
GROUP BY
   rc.kategori,
    rk.total_record
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_record_filter,total_record,rate
Diagnostik,9,16,56.25
Consumable,6,13,46.15
Bedah & Tindakan,5,11,45.45
Perawatan & Rawat Inap,3,11,27.27
Laboratorium,5,8,62.50


**Findings**  
Among the categories reliable for comparison (Diagnostic, Consumable, Surgery & Procedures — Patient Care & Inpatient and Laboratory excluded due to observed and/or expected counts below 5), the proportion of System Input Error is slightly higher in Diagnostic (56.25%, observed 9 vs. expected 7.59; +8.79 points above the 47.46% baseline). Consumable and Surgery & Procedures remain close to baseline (46.15% and 45.45%, respectively).

In [ ]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT (*) AS total_record
FROM
    return_cleaned
WHERE
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY
   customer_type_cleaned
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_record
RS Swasta,11
Klinik,9
RS Pemerintah,8


In [8]:
%%sql

WITH ctc AS (
    SELECT
        customer_type_cleaned,
        COUNT (*) AS total_record
    FROM
        return_cleaned
    WHERE
        return_reason_cleaned = 'Salah Input Sistem'
    GROUP BY
        customer_type_cleaned
)

SELECT
    rc.customer_type_cleaned,
    ctc.total_record AS total_record_filter,
    COUNT (*) AS total_record,
    ROUND((NULLIF(ctc.total_record,0) * 1.0 / NULLIF(COUNT(*), 0))*100, 2) AS rate
FROM
    return_cleaned rc
LEFT JOIN ctc
    ON rc.customer_type_cleaned = ctc.customer_type_cleaned
GROUP BY
   rc.customer_type_cleaned,
   ctc.total_record
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_record_filter,total_record,rate
RS Swasta,11,27,40.74
Klinik,9,19,47.37
RS Pemerintah,8,13,61.54


**Findings**  
All three customer types are reliable for comparison. Public hospitals show the highest proportion of System Input Error (61.54%, observed 8 vs. expected 6.17; +14.08 points above the 47.46% baseline). Clinics are nearly identical to baseline (47.37%), while private hospitals fall slightly below baseline (40.74%, -6.72 points).

In [37]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT(*) FILTER (WHERE kategori = 'Bedah & Tindakan' ) AS Bedah_and_Tindakan,
    COUNT(*) FILTER (WHERE kategori = 'Diagnostik' ) AS Diagnostik,
    COUNT(*) FILTER (WHERE kategori = 'Consumable' ) AS Consumable,
    COUNT(*) FILTER (WHERE kategori = 'Laboratorium' ) AS Laboratorium,
    COUNT(*) FILTER (WHERE kategori = 'Perawatan & Rawat Inap' ) AS Perawatan_and_Rawat_Inap,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY
    customer_type_cleaned
ORDER BY
    customer_type_cleaned

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,bedah_and_tindakan,diagnostik,consumable,laboratorium,perawatan_and_rawat_inap,total_records
Klinik,0,4,3,2,0,9
RS Pemerintah,2,1,2,2,1,8
RS Swasta,3,4,1,1,2,11


In [38]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3
    ) AS Q1,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6
    ) AS Q2,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9
    ) AS Q3,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12
    ) AS Q4,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY
    customer_type_cleaned
ORDER BY
    customer_type_cleaned


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,q1,q2,q3,q4,total_records
Klinik,2,2,2,3,9
RS Pemerintah,0,4,1,3,8
RS Swasta,0,4,3,4,11


**Findings**  
Cross-dimension exploration for System Input Error (quarter × region, quarter × category, quarter × customer type, region × category, region × customer type, category × customer type) shows expected counts below the minimum threshold (≥5) across all cell combinations — a structural consequence of the System Input Error base being only 28 cases, which makes any two-dimensional breakdown too granular to compare reliably. As a result, no cross-dimension pattern can be concluded from this breakdown.

---

In [5]:
%%sql

SELECT
    kategori,
    COUNT(*) AS total_record
FROM
    return_cleaned
GROUP BY
    kategori
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_record
Diagnostik,16
Consumable,13
Bedah & Tindakan,11
Perawatan & Rawat Inap,11
Laboratorium,8


In [6]:
%%sql

WITH
    kategori_total AS (
        SELECT
            COUNT(*) AS total_records
        FROM
            return_cleaned
    )

SELECT
    kategori,
    COUNT(*) AS total_records_filter,
    kt.total_records,
    ROUND((COUNT(*)::numeric / NULLIF(kt.total_records,0)::numeric) * 100,2) AS rate
FROM
    return_cleaned rc
CROSS JOIN 
    kategori_total kt
GROUP BY
    kategori,
    kt.total_records
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_records_filter,total_records,rate
Diagnostik,16,59,27.12
Consumable,13,59,22.03
Perawatan & Rawat Inap,11,59,18.64
Bedah & Tindakan,11,59,18.64
Laboratorium,8,59,13.56


**Findings**  
The Diagnostic category records the highest number of returns (16 of 59 cases, 27.12%), followed by Consumable (13 cases, 22.03%), Patient Care & Inpatient and Surgery & Procedures (11 cases each, 18.64%), and Laboratory (8 cases, 13.56%).

In [40]:
%%sql

WITH 
month AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        COUNT(*) AS total_loss
    FROM
        return_cleaned
    GROUP BY
        quarter
),
month_filter AS (
    SELECT
            CASE
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
            END AS quarter,
            COUNT(*) AS total_loss
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        quarter
)

SELECT
    m.quarter,
    mf.total_loss AS total_loss,
    m.total_loss AS total_loss_all,
    ROUND( (COALESCE(mf.total_loss,0) * 1.0 / NULLIF(m.total_loss,0) * 100)::numeric,2) AS ratio

FROM
    month m
LEFT JOIN
    month_filter mf 
        ON m.quarter = mf.quarter
ORDER BY
    m.quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_loss,total_loss_all,ratio
Q1,3,8,37.50
Q2,4,15,26.67
Q3,4,14,28.57
Q4,5,22,22.73


**Findings**  
Of the 4 quarters, only Q4 has an expected count large enough for a reliable comparison (5.97, out of 22 total cases). The result shows the Diagnostic category is actually slightly below expectation in Q4 (observed 5, 22.73% vs. the 27.12% baseline). Q1, Q2, and Q3 were excluded, as their expected counts fall below 5. As such, there is no solid evidence that Diagnostic-category returns cluster in any particular quarter.

In [44]:
%%sql

WITH kategori_reason AS
(
    SELECT
        return_reason_cleaned,
        COUNT(*) AS total_records
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        return_reason_cleaned
)

SELECT
    rc.return_reason_cleaned,
    kr.total_records AS total_records_filter,
    COUNT(*) AS total_records,
    ROUND((NULLIF(kr.total_records,0) * 1.0 / NULLIF(COUNT(*), 0))*100, 2) AS rate

FROM
    return_cleaned rc
LEFT JOIN
    kategori_reason kr
        ON rc.return_reason_cleaned = kr.return_reason_cleaned
GROUP BY
    rc.return_reason_cleaned,
    total_records_filter
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_records_filter,total_records,rate
Salah Input Sistem,9,28,32.14
Salah Pesan Customer,4,20,20.00
Barang Rusak,2,7,28.57
Kadaluarsa,1,4,25.00


**Findings**  
Within the Diagnostic category overall, only System Input Error is reliable enough for comparison (observed 9, expected 7.59) — the result is slightly above baseline (32.14% vs. 27.12%, a narrow ~5-point gap). The other three reasons (Customer Order Error, Damaged Goods, Expired) were excluded due to observed and/or expected counts below 5 — including Customer Order Error, which initially appeared comparable (expected 5.42) but has an observed count of only 4. No sharp deviation in return reasons was found within the Diagnostic category overall — a contrast to the pattern seen in the top 5 returned products, where System Input Error was far more dominant (53% of 15 cases).

In [46]:
%%sql

WITH kategori_customer AS
(
    SELECT
        customer_type_cleaned,
        COUNT(*) AS total_records
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        customer_type_cleaned 
)

SELECT
    rc.customer_type_cleaned,
    kc.total_records AS total_records_filter,
    COUNT(*) AS total_records,
    ROUND((NULLIF(kc.total_records,0) * 1.0 / NULLIF(COUNT(*), 0))*100, 2) AS rate

FROM
    return_cleaned rc
LEFT JOIN
    kategori_customer kc
        ON rc.customer_type_cleaned = kc.customer_type_cleaned
GROUP BY
    rc.customer_type_cleaned,
    total_records_filter
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_records_filter,total_records,rate
Klinik,7,19,36.84
RS Swasta,6,27,22.22
RS Pemerintah,3,13,23.08


**Findings**  
Clinics show a higher proportion of Diagnostic-category returns than the company average (36.84% vs. a 27.12% baseline, a +9.72 point gap; observed 7 of 19 cases vs. an expected 5.15 under the overall pattern). Private hospitals, by contrast, fall slightly below average (22.22% vs. the 27.12% baseline, a -4.90 point gap; observed 6 of 27 cases vs. an expected 7.32). Public hospitals were excluded from this comparison, as their expected count (3.53) is too small for a reliable comparison.

In [49]:
%%sql

WITH kategori_region AS
(
    SELECT
        region_cleaned,
        COUNT(*) AS total_records
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        region_cleaned
)

SELECT
    rc.region_cleaned,
    kr.total_records AS total_records_filter,
    COUNT(*) AS total_records,
    ROUND((NULLIF(kr.total_records,0) * 1.0 / NULLIF(COUNT(*), 0))*100, 2) AS rate

FROM
    return_cleaned rc
LEFT JOIN
    kategori_region kr
        ON rc.region_cleaned = kr.region_cleaned
GROUP BY
    rc.region_cleaned,
    total_records_filter
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_records_filter,total_records,rate
Bogor,None,4,None
Jakarta Timur,None,6,None
Jakarta Barat,7,13,53.85
Jakarta Selatan,3,10,30.00
Bekasi,2,10,20.00
Depok,1,3,33.33
Jakarta Pusat,1,3,33.33
Jakarta Utara,1,5,20.00
Tangerang,1,5,20.00


**Findings**  
No region meets the minimum expected count threshold (≥5) for a reliable comparison — including Jakarta Barat, which has the highest rate (53.85%, observed 7 vs. expected 3.53). As a result, no region can be concluded as a solid pattern at this stage. *Exploratory note: Jakarta Barat appears higher than expected, but the sample is too small to draw a statistical conclusion — additional data would be needed before this could be treated as a finding*.

In [50]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'RS Swasta' ) AS RS_Swasta,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'Klinik' ) AS Klinik,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'RS Pemerintah' ) AS RS_Pemerintah,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    return_reason_cleaned
ORDER BY
    return_reason_cleaned

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,rs_swasta,klinik,rs_pemerintah,total_records
Barang Rusak,0,0,2,2
Kadaluarsa,1,0,0,1
Salah Input Sistem,4,4,1,9
Salah Pesan Customer,1,3,0,4


In [34]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3
    ) AS Q1,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6
    ) AS Q2,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9
    ) AS Q3,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12
    ) AS Q4,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    customer_type_cleaned
ORDER BY
    customer_type_cleaned


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,q1,q2,q3,q4,total_records
Klinik,2,2,0,3,7
RS Pemerintah,1,0,1,1,3
RS Swasta,0,2,3,1,6


In [35]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3
    ) AS Q1,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6
    ) AS Q2,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9
    ) AS Q3,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12
    ) AS Q4,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    return_reason_cleaned
ORDER BY
    return_reason_cleaned


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,q1,q2,q3,q4,total_records
Barang Rusak,1,0,1,0,2
Kadaluarsa,0,0,1,0,1
Salah Input Sistem,1,3,2,3,9
Salah Pesan Customer,1,1,0,2,4


**Findings**  
Cross-dimension exploration within the Diagnostic category (reason × customer type, customer type × quarter, reason × quarter) all failed to meet the minimum expected count threshold (≥5) in every cell — a structural consequence of the Diagnostic category's small base (16 cases), which makes any two-dimensional breakdown too granular to compare reliably. As a result, no cross-dimension pattern can be concluded from this breakdown.

In [22]:
%%sql

SELECT
    sku_name,
    COUNT(*) AS total_record
FROM
    return_cleaned
GROUP BY
    sku_name
ORDER BY
    total_record DESC
LIMIT 10

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


sku_name,total_record
Urine Stick 10 Parameter,3
Infus Set Anak,3
Selimut Pasien Rumah Sakit,3
Benang Catgut 2-0,3
Syringe 1ml,3
Tensimeter Manual Dewasa,2
Baju Pasien Dewasa,2
Rapid Test HbsAg,2
Duk Operasi 80x90cm,2
NGT 16Fr,2


In [38]:
%%sql

WITH
    sku_total AS (
        SELECT
            COUNT(*) AS total_records
        FROM
            return_cleaned
    )


SELECT
    sku_name,
    COUNT(*) AS total_records_filter,
    kt.total_records,
    ROUND((COUNT(*)::numeric / NULLIF(kt.total_records,0)::numeric) * 100,2) AS rate
FROM
    return_cleaned rc
CROSS JOIN 
    sku_total kt
GROUP BY
    sku_name,
    kt.total_records
ORDER BY
    total_records_filter DESC
LIMIT 10

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


sku_name,total_records_filter,total_records,rate
Syringe 1ml,3,59,5.08
Urine Stick 10 Parameter,3,59,5.08
Infus Set Anak,3,59,5.08
Selimut Pasien Rumah Sakit,3,59,5.08
Benang Catgut 2-0,3,59,5.08
IV Catheter 22G,2,59,3.39
Gunting Bedah Bengkok 14cm,2,59,3.39
NGT 16Fr,2,59,3.39
Tensimeter Manual Dewasa,2,59,3.39
Baju Pasien Dewasa,2,59,3.39


**Findings**  
Finding: Five products are tied for the highest return frequency, each with 3 cases (5.08% of the total 59 returns): Syringe 1ml, 10-Parameter Urine Stick, Pediatric Infusion Set, Hospital Patient Blanket, and 2-0 Catgut Suture. No single product stands out significantly from the rest.

In [22]:
%%sql

SELECT
    sku_name,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'RS Swasta') AS RS_Swasta,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'RS Pemerintah') AS RS_Pemerintah,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'Klinik') AS Klinik,
    COUNT(*) AS total_records
FROM 
    return_cleaned
GROUP BY   
    sku_name
HAVING 
    COUNT(*) = 3
ORDER BY 
    sku_name

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


sku_name,rs_swasta,rs_pemerintah,klinik,total_records
Benang Catgut 2-0,2,1,0,3
Infus Set Anak,1,0,2,3
Selimut Pasien Rumah Sakit,3,0,0,3
Syringe 1ml,1,2,0,3
Urine Stick 10 Parameter,0,0,3,3


**Findings**  
Among the top 5 returned products (n=15 return cases), private hospitals account for the largest share (7 cases, 47%), followed by clinics (5 cases, 33%) and public hospitals (3 cases, 20%). *Note: the small sample size (n=15) means this distribution is indicative rather than a definitive pattern across the broader return population*.

In [ ]:
%%sql

SELECT
    sku_name,
    COUNT(*) FILTER (WHERE return_reason_cleaned = 'Salah Input Sistem') AS Salah_Input_Sistem,
    COUNT(*) FILTER (WHERE return_reason_cleaned = 'Salah Pesan Customer') AS Salah_Pesan_Customer,
    COUNT(*) FILTER (WHERE return_reason_cleaned = 'Barang Rusak') AS Barang_Rusak,
    COUNT(*) FILTER (WHERE return_reason_cleaned = 'Kadaluarsa') AS Kadaluarsa,
    COUNT(*) AS total_records
FROM 
    return_cleaned
GROUP BY   
    sku_name
HAVING 
    COUNT(*) = 3
ORDER BY 
    sku_name


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


sku_name,salah_input_sistem,salah_pesan_customer,barang_rusak,kadaluarsa,total_records
Benang Catgut 2-0,2,1,0,0,3
Infus Set Anak,2,1,0,0,3
Selimut Pasien Rumah Sakit,1,2,0,0,3
Syringe 1ml,2,0,1,0,3
Urine Stick 10 Parameter,1,2,0,0,3


**Findings**  
Among the top 5 returned products (n=15 return cases), the dominant return reason is System Input Error (8 cases, 53%), followed by Customer Order Error (6 cases, 40%), while Damaged Goods accounts for only 1 case (7%). *Note: this is a small sample (n=15, averaging just 3 cases per product), so these proportions are indicative rather than conclusive for the broader return population*.

In [24]:
%%sql

SELECT
    sku_name,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3
    ) AS Q1,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6
    ) AS Q2,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9
    ) AS Q3,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12
    ) AS Q4,
    COUNT(*) AS total_records
FROM
    return_cleaned
GROUP BY
    sku_name
HAVING
    COUNT(*) = 3
ORDER BY
    sku_name;


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


sku_name,q1,q2,q3,q4,total_records
Benang Catgut 2-0,0,2,0,1,3
Infus Set Anak,0,1,1,1,3
Selimut Pasien Rumah Sakit,0,0,1,2,3
Syringe 1ml,0,3,0,0,3
Urine Stick 10 Parameter,1,1,0,1,3


**Findings**  
Among the top 5 returned products (n=15 return cases), Q2 and Q4 account for the majority of cases (7 and 5 cases respectively; 12 of 15, 80% combined), while Q1 and Q3 are notably lower (1 and 2 cases). *Note: this is based on a small sample (n=15) and a single year of data (Jan–Dec 2024), so this pattern is indicative only and cannot yet be confirmed as a recurring seasonal trend*.

In [29]:
%%sql

SELECT
    kategori,
    COUNT(*) AS total_records
FROM
    return_cleaned
GROUP BY
    kategori
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_records
Diagnostik,16
Consumable,13
Bedah & Tindakan,11
Perawatan & Rawat Inap,11
Laboratorium,8


In [ ]:
%%sql

SELECT
    sku_name,
    COUNT(*) FILTER (WHERE kategori = 'Diagnostik') AS Diagnostik,
    COUNT(*) FILTER (WHERE kategori = 'Consumable') AS Consumable,
    COUNT(*) FILTER (WHERE kategori = 'Bedah & Tindakan') AS Bedah_dan_Tindakan,
    COUNT(*) FILTER (WHERE kategori = 'Laboratorium') AS Laboratorium,
    COUNT(*) FILTER (WHERE kategori = 'Perawatan & Rawat Inap') AS Perawatan_dan_Rawat_Inap,
    COUNT(*) AS total_records
FROM 
    return_cleaned
GROUP BY   
    sku_name
HAVING 
    COUNT(*) = 3
ORDER BY 
    sku_name


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


sku_name,diagnostik,consumable,bedah_dan_tindakan,laboratorium,perawatan_dan_rawat_inap,total_records
Benang Catgut 2-0,0,0,3,0,0,3
Infus Set Anak,0,3,0,0,0,3
Selimut Pasien Rumah Sakit,0,0,0,0,3,3
Syringe 1ml,0,3,0,0,0,3
Urine Stick 10 Parameter,3,0,0,0,0,3


**Findings**  
Finding: The top 5 products are distributed across 5 different categories, with no category contributing more than 2 products (Consumable accounts for 2, the rest 1 each). Notably, Diagnostic — the category with the highest overall return count (16 of 59 cases) — is represented by only 1 product in the top 5, contributing just 3 of its 16 cases (18.75%). The remaining 81% is spread across other Diagnostic products outside the top 5.

In [26]:
%%sql

SELECT
    sku_name,
    COUNT(*) FILTER (WHERE region_cleaned = 'Bekasi') AS Bekasi,
    COUNT(*) FILTER (WHERE region_cleaned = 'Bogor') AS Bogor,
    COUNT(*) FILTER (WHERE region_cleaned = 'Depok') AS Depok,
    COUNT(*) FILTER (WHERE region_cleaned = 'Jakarta Barat') AS Jakarta_Barat,
    COUNT(*) FILTER (WHERE region_cleaned = 'Jakarta Pusat') AS Jakarta_Pusat,
    COUNT(*) FILTER (WHERE region_cleaned = 'Jakarta Selatan') AS Jakarta_selatan,
    COUNT(*) FILTER (WHERE region_cleaned = 'Jakarta Timur') AS Jakarta_timur,
    COUNT(*) FILTER (WHERE region_cleaned = 'Jakarta Utara') AS Jakarta_utara,
    COUNT(*) FILTER (WHERE region_cleaned = 'Tangerang') AS Tangerang,
    COUNT(*) AS total_records
FROM 
    return_cleaned
GROUP BY   
    sku_name
HAVING 
    COUNT(*) = 3
ORDER BY 
    sku_name

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


sku_name,bekasi,bogor,depok,jakarta_barat,jakarta_pusat,jakarta_selatan,jakarta_timur,jakarta_utara,tangerang,total_records
Benang Catgut 2-0,1,0,0,0,0,0,0,2,0,3
Infus Set Anak,0,1,0,1,0,0,1,0,0,3
Selimut Pasien Rumah Sakit,1,0,0,0,0,2,0,0,0,3
Syringe 1ml,0,0,0,0,0,2,0,0,1,3
Urine Stick 10 Parameter,2,0,0,0,0,1,0,0,0,3


**Findings**  
Finding: Among the top 5 returned products (n=15 return cases), the regional distribution does not show a pattern strong enough to draw conclusions — cases are spread across 9 different regions with very small counts per cell (highest is only 5 cases), making it easy to misread as a pattern when it may simply be chance.

**Reason's Insight**  
Among all reliable breakdowns, the strongest signal for System Input Error appears in the April–June 2024 period (Q2), which spikes notably above baseline (+19.21 points) — far sharper than the more moderate signals seen in category, region, and customer type (Diagnostic, Jakarta Barat, and public hospitals, ranging +8.79 to +14.08 points). However, whether these signals are actually connected — for instance, whether the Q2 spike is largely driven by public hospitals or Jakarta Barat — cannot be confirmed, as every cross-dimension combination failed the reliability threshold once split across two dimensions. It's also worth noting that the available data covers only a single year (2024), so this Q2 spike reflects a specific period rather than a seasonal pattern confirmed to recur annually. As such, the April–June 2024 period stands out as the strongest statistically-grounded starting point for investigation within the available data, while the patterns in category, region, and customer type should be treated as early hypotheses requiring further validation.

**Category's Insight**  
Among all reliable breakdowns within the Diagnostic category, the strongest signal appears in the customer type dimension — clinics are significantly over-represented above baseline (+9.72 points), sharper than the thin signal seen in return reason (System Input Error, +5.02 points) or the time dimension, which actually shows the opposite direction (Q4 slightly below expectation, -4.39 points, with no evidence of quarterly clustering). Region cannot be concluded at all, as every cell failed the reliability threshold — including Jakarta Barat, which appears elevated (53.85% rate) but with too small a sample to be treated as a pattern. Attempts to connect the clinic signal to a specific reason — for instance, testing whether clinics' over-representation is driven by System Input Error — could not be confirmed, as the reason × customer type combination also failed the reliability threshold. As such, customer type (clinics) is the strongest statistically-grounded dimension for further investigation into elevated Diagnostic-category returns, while reason, time, and region cannot yet support a solid conclusion.

**Product's Insight**  
At the individual product level, none of the breakdowns (customer segment, reason, time, region) is reliable enough to support a solid pattern — all are based on n=15 (top 5 products), too small to clear the statistical reliability threshold. However, the consistent absence of a pattern is itself informative: returns at the product level appear diffusely spread rather than concentrated. This is reinforced by the category breakdown (Finding #5) — the top 5 products are spread across 5 different categories, and even Diagnostic, the category with the highest overall return volume (16 of 59 cases), is represented by only 1 product in the top 5 (contributing just 18.75% of Diagnostic's total cases). This suggests that Diagnostic's elevated return volume is not driven by one or two problematic SKUs, but by many different products each contributing a small share. As such, a "find the problem product" strategy is unlikely to be productive at this level — investigation and intervention would be more effective at the category or customer-type level (such as clinics, which showed a reliable signal in the category-level analysis), rather than at the individual SKU level.

**Reason's Conclusion:** 
 The most common cause of product returns is System Input Error, accounting for nearly half of all return cases (28 of 59, 47.46%) — well above Customer Order Error (33.90%), Damaged Goods (11.86%), and Expired (6.78%). This finding is based on the full population (n=59), making it a reliable primary answer.

On closer examination, System Input Error's share is most pronounced during the April–June 2024 period (+19.21 points above baseline), with more moderate secondary signals in the Diagnostic category, Jakarta Barat region, and public hospital customer type (+8.79 to +14.08 points each). However, the causal relationship between these signals could not be confirmed — every cross-dimension combination failed the reliability threshold due to sample limitations. The available data also spans only one year (2024), so the April–June spike cannot yet be confirmed as a recurring seasonal pattern.

As such, System Input Error stands as the primary, data-backed intervention target, with the April–June 2024 period as the top-priority starting point for operational root-cause investigation — while the patterns across category, region, and customer type remain hypotheses requiring further data before informing decisions.

**Category and Products's Conclusion:** 
At the category level, returns occur most frequently within the Diagnostic category (16 of 59 cases, 27.12%) — this finding is based on the full population (n=59), making it a reliable primary answer. However, at the individual product level, no single product stands out — five products (Syringe 1ml, 10-Parameter Urine Stick, Pediatric Infusion Set, Hospital Patient Blanket, 2-0 Catgut Suture) are tied at the top with identical frequency (3 cases each).

On closer examination, Diagnostic's elevated return volume is not driven by one or two problematic SKUs — the top 5 products are in fact spread across 5 different categories, and even the Diagnostic-category product among them accounts for only 18.75% of Diagnostic's total cases. A more reliable signal instead appears in the customer type dimension: clinics are significantly over-represented within Diagnostic-category returns (+9.72 points above baseline, reliable), making this the strongest dimension for further investigation, rather than searching for a specific problem SKU.

As such, the answer to this question differs by level: at the category level, Diagnostic is a clear, well-supported answer; at the product level, no single reliable answer exists — the data is too thinly spread to draw a conclusion. Investigation and intervention should be directed toward the combination of category (Diagnostic) and customer type (clinics), rather than toward identifying a single "most problematic" SKU.

# 2. Financial Impact

In [64]:
%%sql

SELECT
    SUM(value_return) AS total_loss,
    AVG(value_return) AS avg
FROM
    return_cleaned


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_loss,avg
171506500.0,2906889.8305084747


**Finding**  
Total financial loss from returns throughout 2024 reached IDR 171,506,500 across 59 return cases, with an average loss per case of IDR 2,906,890.

In [65]:
%%sql

SELECT
    kategori,
    SUM(value_return) AS total_loss,
    COUNT(value_return) AS total_record
FROM
    return_cleaned
GROUP BY
    kategori
ORDER BY
    total_loss DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_loss,total_record
Diagnostik,66811000.0,16
Bedah & Tindakan,53336500.0,11
Perawatan & Rawat Inap,22706500.0,11
Consumable,21474000.0,13
Laboratorium,7178500.0,8


In [4]:
%%sql

WITH
    kategori_total AS (
        SELECT
            SUM(value_return) AS total_loss
        FROM
            return_cleaned
    )

SELECT
    kategori,
    SUM(value_return) AS total_loss_filter,
    kt.total_loss,
    ROUND((SUM(value_return)::numeric / NULLIF(kt.total_loss,0)::numeric) * 100,2) AS rate,
    COUNT(value_return) AS total_record_filter,
    AVG(value_return) AS average_loss_filter
FROM
    return_cleaned rc
CROSS JOIN 
    kategori_total kt
GROUP BY
    kategori,
    kt.total_loss
ORDER BY
    total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_loss_filter,total_loss,rate,total_record_filter,average_loss_filter
Diagnostik,66811000.0,171506500.0,38.96,16,4175687.5
Bedah & Tindakan,53336500.0,171506500.0,31.10,11,4848772.7272727275
Perawatan & Rawat Inap,22706500.0,171506500.0,13.24,11,2064227.2727272727
Consumable,21474000.0,171506500.0,12.52,13,1651846.1538461538
Laboratorium,7178500.0,171506500.0,4.19,8,897312.5


**Finding**  
Diagnostic is the category with the highest total return loss (IDR 66,811,000 out of IDR 171,506,500, or 38.96%, from 16 cases), with an average loss per case of around IDR 4,175,688. Interestingly, when looked at by average loss per case, Surgery & Procedures actually has the highest average (IDR 4,848,773) despite having fewer cases (11 of 59) and a lower total (IDR 53,336,500, 31.10%).

In [12]:
%%sql

WITH 
month AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        SUM(value_return) AS total_loss,
        COUNT(*) AS total_records
    FROM
        return_cleaned
    GROUP BY
        quarter
),
month_filter AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        SUM(value_return) AS total_loss,
        COUNT(*) AS total_record_filter,
        AVG(value_return) AS average_loss_filter
    FROM 
        return_cleaned
    WHERE 
        kategori = 'Diagnostik'
    GROUP BY 
        quarter
)


SELECT
    m.quarter,
    COALESCE(mf.total_loss,0) AS total_loss,
    m.total_loss AS total_loss_all,
    ROUND((COALESCE(mf.total_loss,0) * 1.0 / NULLIF(m.total_loss,0) * 100)::numeric,2) AS rate,
    COALESCE(mf.total_record_filter,0) AS total_record_filter,
    COALESCE(m.total_records,0) AS total_records,
    COALESCE(mf.average_loss_filter,0) AS average_loss_filter
FROM month m
LEFT JOIN month_filter mf 
    ON m.quarter = mf.quarter
ORDER BY 
    m.quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_loss,total_loss_all,rate,total_record_filter,total_records,average_loss_filter
Q1,3510000.0,30449500.0,11.53,3,8,1170000.0
Q2,33395000.0,69647000.0,47.95,4,15,8348750.0
Q3,25298000.0,34531500.0,73.26,4,14,6324500.0
Q4,4608000.0,36878500.0,12.50,5,22,921600.0


**Findings**
Of the 4 quarters, only Q4 is reliable enough for comparison (observed 5 cases, expected 5.97 cases, narrowly above the threshold). The result shows Diagnostic-category loss in Q4 falls well below expectation (deviation of -IDR 9,758,168; rate 12.50% vs. the 38.96% baseline). In contrast, Q3 shows a high positive deviation (+IDR 11,846,116, 73.26% rate) — but this quarter is not reliable, as its case count (4) falls below the threshold.

In [31]:
%%sql

WITH kategori_reason AS
(
    SELECT
        return_reason_cleaned,
        COUNT(*) AS total_records_filter,
        SUM(value_return) AS total_loss_filter,
        AVG(value_return) AS average_loss
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        return_reason_cleaned
)

SELECT
    rc.return_reason_cleaned,
    kr.total_loss_filter AS total_loss_filter,
    SUM(rc.value_return) AS total_loss,
    ROUND((COALESCE(kr.total_loss_filter,0) * 1.0 / NULLIF(SUM(rc.value_return), 0)*100)::numeric, 2) AS rate,
    ROUND(kr.average_loss::numeric, 2) AS average_loss_filter,
    kr.total_records_filter AS total_records_filter,
    COUNT(*) AS total_records
    
FROM
    return_cleaned rc
LEFT JOIN
    kategori_reason kr
        ON rc.return_reason_cleaned = kr.return_reason_cleaned
GROUP BY
    rc.return_reason_cleaned,
    total_loss_filter,
    total_records_filter,
    average_loss_filter
ORDER BY
    total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_loss_filter,total_loss,rate,average_loss_filter,total_records_filter,total_records
Salah Input Sistem,26842000.0,66634500.0,40.28,2982444.44,9,28
Salah Pesan Customer,19761000.0,70452500.0,28.05,4940250.00,4,20
Barang Rusak,17447000.0,29144000.0,59.86,8723500.00,2,7
Kadaluarsa,2761000.0,5275500.0,52.34,2761000.00,1,4


**Findings**  
System Input Error is the only reliable reason for Diagnostic-category loss (observed 9 cases vs. expected 7.59, both above threshold), with a loss of IDR 26,842,000 and a rate of 40.28% — slightly above (+1.32 points) the 38.96% baseline, with a deviation of +IDR 881,199. Customer Order Error and Damaged Goods are both unreliable, though from different angles: Customer Order Error has a passing expected count (5.42) but only 4 observed cases, while Damaged Goods records the highest rate among all reasons (59.86%) but is based on just 2 of 7 cases — both too small to support a conclusion.

In [33]:
%%sql

WITH kategori_customer AS
(
    SELECT
        customer_type_cleaned,
        COUNT(*) AS total_records,
        SUM(value_return) AS total_loss,
        AVG(value_return) AS average_loss
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        customer_type_cleaned 
)

SELECT
    rc.customer_type_cleaned,
    kc.total_loss AS total_loss_filter,
    SUM(rc.value_return) AS total_loss,
    ROUND((COALESCE(kc.total_loss,0) * 1.0 / NULLIF(SUM(rc.value_return), 0)*100)::numeric, 2) AS rate,
    ROUND(kc.average_loss::numeric, 2) AS average_loss_filter,
    kc.total_records AS total_records_filter,
    COUNT(*) AS total_records
FROM
    return_cleaned rc
LEFT JOIN
    kategori_customer kc
        ON rc.customer_type_cleaned = kc.customer_type_cleaned
GROUP BY
    rc.customer_type_cleaned,
    total_loss_filter,
    total_records_filter,
    average_loss_filter
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_loss_filter,total_loss,rate,average_loss_filter,total_records_filter,total_records
Klinik,29225000.0,52240500.0,55.94,4175000.00,7,19
RS Swasta,18305000.0,76510000.0,23.92,3050833.33,6,27
RS Pemerintah,19281000.0,42756000.0,45.10,6427000.00,3,13


**Findings**  
Clinics are the only customer type reliable across all aspects for Diagnostic-category loss — observed 7 cases vs. expected 5.15 cases (both above threshold), with a rate of 55.94% (+16.98 points above the 38.96% baseline) and a deviation of +IDR 8,872,101. Private hospitals are also reliable by case count, but their rate falls below baseline (23.92%). In contrast, public hospitals record the highest rate (45.10%) but are not reliable at all, with a base of only 3 of 13 cases.

In [35]:
%%sql

WITH kategori_region AS
(
    SELECT
        region_cleaned,
        COUNT(*) AS total_records,
        SUM(value_return) AS total_loss,
        AVG(value_return) AS average_loss
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        region_cleaned
)

SELECT
    rc.region_cleaned,
    kr.total_loss AS total_loss_filter,
    SUM(rc.value_return) AS total_loss,
    ROUND((COALESCE(kr.total_loss,0) * 1.0 / NULLIF(SUM(rc.value_return), 0)*100)::numeric, 2) AS rate,
    ROUND((COALESCE(kr.average_loss,0))::numeric, 2) AS average_loss_filter,
    kr.total_records AS total_records_filter,
    COUNT(*) AS total_records
FROM
    return_cleaned rc
LEFT JOIN
    kategori_region kr
        ON rc.region_cleaned = kr.region_cleaned
GROUP BY
    rc.region_cleaned,
    total_records_filter,
    total_loss_filter,
    average_loss_filter
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_loss_filter,total_loss,rate,average_loss_filter,total_records_filter,total_records
Bogor,None,21657000.0,0.00,0.00,None,4
Jakarta Timur,None,4650500.0,0.00,0.00,None,6
Jakarta Barat,19953000.0,33920000.0,58.82,2850428.57,7,13
Jakarta Selatan,7182000.0,11695500.0,61.41,2394000.00,3,10
Bekasi,18734000.0,36964500.0,50.68,9367000.00,2,10
Depok,1908000.0,3399500.0,56.13,1908000.00,1,3
Jakarta Pusat,1587000.0,19224000.0,8.26,1587000.00,1,3
Jakarta Utara,1468000.0,19318500.0,7.60,1468000.00,1,5
Tangerang,15979000.0,20677000.0,77.28,15979000.00,1,5


**Findings**  
No region meets the reliability threshold, primarily due to expected count — even Jakarta Barat, whose observed count passes (7 cases, above the threshold of 5), still fails on expected count (3.53). Exploratory note: Jakarta Barat records a 58.82% rate (+19.86 points above the 38.96% baseline, a deviation of +IDR 6,737,608), and Tangerang shows the highest rate among all regions (77.28%, a deviation of +IDR 7,923,223) — but Tangerang's base is only 1 of 5 cases, well below threshold. Neither figure supports a solid conclusion.

In [38]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'RS Swasta' ) AS RS_Swasta,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'Klinik' ) AS Klinik,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'RS Pemerintah' ) AS RS_Pemerintah,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    return_reason_cleaned
ORDER BY
    return_reason_cleaned

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,rs_swasta,klinik,rs_pemerintah,total_records
Barang Rusak,0,0,2,2
Kadaluarsa,1,0,0,1
Salah Input Sistem,4,4,1,9
Salah Pesan Customer,1,3,0,4


In [44]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3
    ) AS Q1,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6
    ) AS Q2,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9
    ) AS Q3,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12
    ) AS Q4,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    return_reason_cleaned
ORDER BY
    return_reason_cleaned


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,q1,q2,q3,q4,total_records
Barang Rusak,1,0,1,0,2
Kadaluarsa,0,0,1,0,1
Salah Input Sistem,1,3,2,3,9
Salah Pesan Customer,1,1,0,2,4


In [45]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3
    ) AS Q1,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6
    ) AS Q2,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9
    ) AS Q3,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12
    ) AS Q4,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    customer_type_cleaned
ORDER BY
    customer_type_cleaned


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,q1,q2,q3,q4,total_records
Klinik,2,2,0,3,7
RS Pemerintah,1,0,1,1,3
RS Swasta,0,2,3,1,6


**Findings**  
As with the frequency analysis (Section 1), cross-dimension exploration within the Diagnostic category (reason × customer type, customer type × quarter, reason × quarter) fails the minimum reliability threshold across all cells examined — a structural consequence of the Diagnostic category's small base (16 cases), which makes any two-dimensional split too granular to compare reliably. As such, no cross-dimension combination can support a solid financial loss conclusion.

---

In [48]:
%%sql

SELECT
    sku_name,
    SUM(value_return) AS total_loss,
    COUNT(value_return) AS total_record
FROM
    return_cleaned
GROUP BY
    sku_name
ORDER BY
    total_loss DESC
LIMIT 10

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


sku_name,total_loss,total_record
Rapid Test HbsAg,24174000.0,2
Urine Stick 10 Parameter,19189000.0,3
Benang Catgut 2-0,17976000.0,3
IV Catheter 22G,16345500.0,2
Handscoon Steril 6.5,15645000.0,1
Handscoon Steril 7.5,12801000.0,1
Kursi Roda Standar,9219000.0,1
Glukometer Set,7980000.0,2
Tiang Infus Stainless 5-Roda,6207500.0,1
Pulse Oximeter,5077000.0,2


In [56]:
%%sql

WITH
    sku_total AS (
        SELECT
            SUM(value_return) AS total_loss
        FROM
            return_cleaned
    )

SELECT
    sku_name,
    SUM(value_return) AS total_loss,
    sku.total_loss AS grand_total_loss,
    ROUND((SUM(value_return)::numeric / NULLIF(sku.total_loss,0)::numeric) * 100,2) AS rate,
    COUNT(value_return) AS total_record_filter,
    ROUND(AVG(value_return):: numeric,2) AS average_loss_filter
FROM
    return_cleaned rc
CROSS JOIN 
    sku_total sku
GROUP BY
    sku_name,
    sku.total_loss
ORDER BY
    total_loss DESC
LIMIT 10

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


sku_name,total_loss,grand_total_loss,rate,total_record_filter,average_loss_filter
Rapid Test HbsAg,24174000.0,171506500.0,14.10,2,12087000.00
Urine Stick 10 Parameter,19189000.0,171506500.0,11.19,3,6396333.33
Benang Catgut 2-0,17976000.0,171506500.0,10.48,3,5992000.00
IV Catheter 22G,16345500.0,171506500.0,9.53,2,8172750.00
Handscoon Steril 6.5,15645000.0,171506500.0,9.12,1,15645000.00
Handscoon Steril 7.5,12801000.0,171506500.0,7.46,1,12801000.00
Kursi Roda Standar,9219000.0,171506500.0,5.38,1,9219000.00
Glukometer Set,7980000.0,171506500.0,4.65,2,3990000.00
Tiang Infus Stainless 5-Roda,6207500.0,171506500.0,3.62,1,6207500.00
Pulse Oximeter,5077000.0,171506500.0,2.96,2,2538500.00


**Findings**  
The top 10 products by loss are each based on just 1–3 return cases (out of 59 total) — Rapid Test HbsAg records the highest loss (IDR 24,174,000, 14.10% of total, from 2 cases), followed by 10-Parameter Urine Stick (IDR 19,189,000, 11.19%, 3 cases) and 2-0 Catgut Suture (IDR 17,976,000, 10.48%, 3 cases). Given the very small case base per product, further breakdown (by time, region, or customer type) is not viable — this mirrors the pattern found at the product level in the frequency analysis (Section 1), where losses are thinly spread across many different products rather than concentrated in a few SKUs.

**Conclusion:** 
 

# 3. Who & Where

In [39]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
GROUP BY
    customer_type_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_records
RS Swasta,27
Klinik,19
RS Pemerintah,13


In [50]:
%%sql

WITH customer_type AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.customer_type_cleaned,
    COUNT(*) AS total_records__filter,
    ct.total_records AS total_records,
    ROUND((COUNT(*)::numeric / NULLIF(ct.total_records,0)::numeric) * 100,2) AS ratio
FROM
    return_cleaned rc
CROSS JOIN 
    customer_type ct
GROUP BY
    rc.customer_type_cleaned,
    ct.total_records
ORDER BY
    total_records__filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_records__filter,total_records,ratio
RS Swasta,27,59,45.76
Klinik,19,59,32.20
RS Pemerintah,13,59,22.03


In [51]:
%%sql

SELECT
    CASE
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    quarter
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_records
Q4,10
Q2,8
Q3,8
Q1,1


In [59]:
%%sql

WITH 
ct_month AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        COUNT(*) AS total_records
    FROM
        return_cleaned
    GROUP BY
        quarter
),
ct_month_filter AS (
    SELECT
            CASE
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
            END AS quarter,
            COUNT(*) AS total_records_filter
    FROM
        return_cleaned
    WHERE
        customer_type_cleaned ='RS Swasta'
    GROUP BY
        quarter
)

SELECT
    ctm.quarter,
    ctmf.total_records_filter,
    ctm.total_records,
    ROUND((ctmf.total_records_filter::numeric / NULLIF(ctm.total_records,0)::numeric) * 100,2) AS ratio

FROM
    ct_month ctm
LEFT JOIN
    ct_month_filter ctmf 
        ON ctm.quarter = ctmf.quarter
ORDER BY
    ctmf.total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_records_filter,total_records,ratio
Q4,10,22,45.45
Q3,8,14,57.14
Q2,8,15,53.33
Q1,1,8,12.50


In [60]:
%%sql

SELECT
    kategori,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    kategori
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_records
Bedah & Tindakan,7
Perawatan & Rawat Inap,7
Diagnostik,6
Laboratorium,4
Consumable,3


In [65]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    return_reason_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_records
Salah Input Sistem,11
Salah Pesan Customer,10
Barang Rusak,4
Kadaluarsa,2


In [66]:
%%sql

WITH ct_kategori_filter AS
(
    SELECT
        return_reason_cleaned,
        COUNT(*) AS total_records
    FROM
        return_cleaned
    WHERE
        customer_type_cleaned ='RS Swasta'
    GROUP BY
        return_reason_cleaned
)

SELECT
    rc.return_reason_cleaned,
    ctkf.total_records AS total_records_filter,
    COUNT(*) AS total_records,
    ROUND((ctkf.total_records::numeric / NULLIF(COUNT(*),0)::numeric) * 100,2) AS ratio
FROM
    return_cleaned rc
LEFT JOIN
    ct_kategori_filter ctkf
        ON rc.return_reason_cleaned = ctkf.return_reason_cleaned
GROUP BY
    rc.return_reason_cleaned,
    total_records_filter
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_records_filter,total_records,ratio
Salah Input Sistem,11,28,39.29
Salah Pesan Customer,10,20,50.00
Barang Rusak,4,7,57.14
Kadaluarsa,2,4,50.00


In [69]:
%%sql

SELECT
    region_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    region_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
8 rows affected.


region_cleaned,total_records
Jakarta Barat,6
Jakarta Selatan,5
Tangerang,4
Bekasi,4
Jakarta Utara,3
Depok,2
Jakarta Timur,2
Jakarta Pusat,1


In [68]:
%%sql

SELECT *
FROM
    return_cleaned
LIMIT 3

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


retur_id,invoice_id,sku_id,customer_id,return_date,sku_name,kategori,value_return,return_reason_cleaned,customer_type_cleaned,region_cleaned,upper_bound
RT00042,010029,D001,RSP053,2024-03-02,Tensimeter Digital Lengan,Diagnostik,1468000.0,Barang Rusak,RS Pemerintah,Jakarta Utara,4107875.0
RT00046,010092,B002,RSS044,2024-05-14,Gunting Bedah Bengkok 14cm,Bedah & Tindakan,417500.0,Salah Pesan Customer,RS Swasta,Jakarta Selatan,4107875.0
RT00021,009825,L002,RSS056,2024-11-22,Tabung Vacutainer Plain,Laboratorium,478000.0,Salah Input Sistem,RS Swasta,Bekasi,4107875.0


In [70]:
%%sql

SELECT
    region_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
GROUP BY
    region_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_records
Jakarta Barat,13
Bekasi,10
Jakarta Selatan,10
Jakarta Timur,6
Tangerang,5
Jakarta Utara,5
Bogor,4
Jakarta Pusat,3
Depok,3


In [ ]:
%%sql

WITH region AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.region_cleaned,
    COUNT(*) AS total_records__filter,
    g.total_records AS total_records,
    ROUND((COUNT(*)::numeric / NULLIF(g.total_records,0)::numeric) * 100,2) AS rate
    return_cleaned rc
CROSS JOIN 
    region g
GROUP BY
    rc.region_cleaned,
    g.total_records
ORDER BY
    total_records__filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_records__filter,total_records,ratio
Jakarta Barat,13,59,22.03
Bekasi,10,59,16.95
Jakarta Selatan,10,59,16.95
Jakarta Timur,6,59,10.17
Jakarta Utara,5,59,8.47
Tangerang,5,59,8.47
Bogor,4,59,6.78
Depok,3,59,5.08
Jakarta Pusat,3,59,5.08


**Conclusion:** 
 

# 4. Trend Over Time

**Conclusion:** 
 